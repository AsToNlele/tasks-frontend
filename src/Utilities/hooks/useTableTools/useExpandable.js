import React, { useState } from 'react';
import { hasDetails } from '../../../SmartComponents/completedTaskDetailsHelpers';

const buildExpandableRows = (
  item,
  row,
  rowIndex,
  options,
  columns,
  isOpen,
  parentIndex
) => {
  let newRow;

  if (hasDetails(item)) {
    newRow = {
      ...row,
      isOpen,
    };

    const expandableRow = itemDetailsRow(
      item,
      rowIndex,
      options,
      columns,
      parentIndex
    );

    return [newRow, expandableRow];
  } else {
    return [row];
  }
};

const calculateColSpan = (columns, options) => {
  let colSpan = columns.length + 1;
  if (options.selectable || options.hasRadioSelect) {
    colSpan++;
  }
  return colSpan;
};

const childRowForJob = (item, idx, DetailsComponent, colSpan, parentIndex) => ({
  parent: parentIndex,
  fullWidth: true,
  cells: [
    {
      title: <DetailsComponent item={item} key={'item-' + item.system} />,
      props: { colSpan },
    },
  ],
});

const itemDetailsRow = (item, idx, options, columns, parentIndex) =>
  typeof options?.detailsComponent !== 'undefined' &&
  childRowForJob(
    item,
    idx,
    options.detailsComponent,
    calculateColSpan(columns, options),
    parentIndex
  );

const useExpandable = (options) => {
  const enableExpandable = !!options.detailsComponent;
  const [openItems, setOpenItems] = useState([]);
  const onCollapse = (_event, _index, _isOpen, row) => {
    if (openItems.includes(row.itemId)) {
      setOpenItems(openItems.filter((itemId) => itemId !== row.itemId));
    } else {
      setOpenItems([...openItems, row.itemId]);
    }
  };

  const openItem = (row, item, columns, rowIndex, parentIndex) => {
    const isOpen = openItems.includes(item.system);
    let newRow = buildExpandableRows(
      item,
      row,
      rowIndex,
      options,
      columns,
      isOpen,
      parentIndex
    );

    return newRow;
  };

  return enableExpandable
    ? {
        transformer: openItem,
        tableProps: {
          onCollapse,
        },
      }
    : {};
};

export default useExpandable;
