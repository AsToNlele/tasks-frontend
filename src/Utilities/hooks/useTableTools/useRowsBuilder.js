import React from 'react';
import moment from 'moment';

const renderRunDateTime = (end) => {
  return end === 'null'
    ? 'Running'
    : moment.utc(end).format('DD MMM YYYY, HH:mm UTC');
};

const buildRow = (item, index) => [
  <div key={`task-title-${index}`}>
    <a href="beta/insights/tasks">{item.title}</a>
  </div>,
  item.system_count,
  renderRunDateTime(item.end),
];

const useRowsBuilder = (items, options = {}) => {
  const filteredItems = options?.filter ? options.filter(items) : items;

  const sortedItems = options?.sorter
    ? options.sorter(filteredItems)
    : filteredItems;

  const paginatedItems = options?.paginator
    ? options?.paginator(filteredItems)
    : sortedItems;

  const rows = paginatedItems.map((item, index) => buildRow(item, index));

  const pagination = options?.pagination
    ? {
        ...options.pagination,
        itemCount: filteredItems.length,
      }
    : undefined;

  return {
    tableProps: { rows },
    toolbarProps: {
      pagination,
    },
  };
};

export default useRowsBuilder;
