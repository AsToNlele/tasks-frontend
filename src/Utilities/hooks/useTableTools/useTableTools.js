import useRowsBuilder from './useRowsBuilder';
import useFilterConfig from './useFilterConfig';
import usePaginate from './usePaginate';
import { useExportWithItems } from './useExport';
import { useTableSortWithItems } from './useTableSort';

const filteredAndSortedItems = (items, filter, sorter) => {
  const filtered = filter ? filter(items) : items;
  return sorter ? sorter(filtered) : filtered;
};

const useTableTools = (items = [], columns = [], options = {}) => {
  const {
    toolbarProps: paginationToolbarProps,
    setPage,
    paginator,
  } = usePaginate(options);

  const {
    toolbarProps: conditionalFilterProps,
    filter,
    selectedFilterToolbarProps,
  } = useFilterConfig({
    ...options,
    setPage,
  });

  const { tableProps: sortableTableProps, sorter } = useTableSortWithItems(
    items,
    columns,
    options
  );

  const { toolbarProps: exportToolbarProps } = useExportWithItems(
    filteredAndSortedItems(items, filter, sorter),
    columns,
    options
  );

  const {
    toolbarProps: rowBuilderToolbarProps,
    tableProps: rowBuilderTableProps,
  } = useRowsBuilder(items, columns, {
    emptyRows: options.tableProps.emptyRows,
    pagination: paginationToolbarProps?.pagination,
    paginator,
    filter,
    sorter,
  });

  const toolbarProps = {
    ...paginationToolbarProps,
    ...conditionalFilterProps,
    ...selectedFilterToolbarProps,
    ...rowBuilderToolbarProps,
    ...exportToolbarProps,
  };

  const tableProps = {
    cells: columns,
    ...rowBuilderTableProps,
    ...sortableTableProps,
  };

  return {
    toolbarProps,
    tableProps,
  };
};

export default useTableTools;
