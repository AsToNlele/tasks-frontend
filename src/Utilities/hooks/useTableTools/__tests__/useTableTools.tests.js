import { renderHook } from '@testing-library/react-hooks';
import useTableTools from '../useTableTools';
import items from './__fixtures__/items.fixtures';
import columns from './__fixtures__/columns.fixtures';

describe('useTableTools', () => {
  it('returns tableProps', () => {
    const { result } = renderHook(() => useTableTools(items, columns));
    expect(result).toMatchSnapshot();
    expect(result.current.tableProps.rows.length).toBe(2);
    expect(result.current.tableProps.cells.length).toBe(3);
  });

  it('should set default with no params', () => {
    const { result } = renderHook(() => useTableTools(undefined, undefined));
    expect(result.current.tableProps.rows.length).toBe(0);
    expect(result.current.tableProps.cells.length).toBe(0);
  });
});
