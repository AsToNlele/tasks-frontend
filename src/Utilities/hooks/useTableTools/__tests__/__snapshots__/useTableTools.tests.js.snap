// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTableTools returns tableProps 1`] = `
Object {
  "all": Array [
    Object {
      "tableProps": Object {
        "cells": Array [
          Object {
            "props": Object {
              "width": 35,
            },
            "sortByProp": "title",
            "title": "Task",
            "transforms": Array [
              [Function],
            ],
          },
          Object {
            "props": Object {
              "width": 20,
            },
            "renderExport": [Function],
            "sortByProp": "system_count",
            "title": "Systems",
            "transforms": Array [
              [Function],
            ],
          },
          Object {
            "props": Object {
              "width": 20,
            },
            "sortByProp": "run_date_time",
            "title": "Run date/time",
            "transforms": Array [
              [Function],
            ],
          },
        ],
        "onSort": [Function],
        "rows": Array [
          Array [
            Object {
              "title": "taskA",
            },
            Object {
              "title": 10,
            },
            Object {
              "title": "Running",
            },
          ],
          Array [
            Object {
              "title": "taskB",
            },
            Object {
              "title": 5,
            },
            Object {
              "title": "20 Apr 2022, 11:10 UTC",
            },
          ],
          Array [
            Object {
              "title": "taskB",
            },
            Object {
              "title": 3,
            },
            Object {
              "title": "25 Apr 2022, 11:10 UTC",
            },
          ],
        ],
        "sortBy": Object {
          "direction": "asc",
          "index": 0,
        },
      },
      "toolbarProps": Object {
        "pagination": Object {
          "itemCount": 3,
          "onPerPageSelect": [Function],
          "onSetPage": [Function],
          "page": 1,
          "perPage": 10,
        },
      },
    },
    Object {
      "tableProps": Object {
        "cells": Array [
          Object {
            "props": Object {
              "width": 35,
            },
            "sortByProp": "title",
            "title": "Task",
            "transforms": Array [
              [Function],
            ],
          },
          Object {
            "props": Object {
              "width": 20,
            },
            "renderExport": [Function],
            "sortByProp": "system_count",
            "title": "Systems",
            "transforms": Array [
              [Function],
            ],
          },
          Object {
            "props": Object {
              "width": 20,
            },
            "sortByProp": "run_date_time",
            "title": "Run date/time",
            "transforms": Array [
              [Function],
            ],
          },
        ],
        "onSort": [Function],
        "rows": Array [
          Array [
            Object {
              "title": "taskA",
            },
            Object {
              "title": 10,
            },
            Object {
              "title": "Running",
            },
          ],
          Array [
            Object {
              "title": "taskB",
            },
            Object {
              "title": 5,
            },
            Object {
              "title": "20 Apr 2022, 11:10 UTC",
            },
          ],
          Array [
            Object {
              "title": "taskB",
            },
            Object {
              "title": 3,
            },
            Object {
              "title": "25 Apr 2022, 11:10 UTC",
            },
          ],
        ],
        "sortBy": Object {
          "direction": "asc",
          "index": 0,
        },
      },
      "toolbarProps": Object {
        "pagination": Object {
          "itemCount": 3,
          "onPerPageSelect": [Function],
          "onSetPage": [Function],
          "page": 1,
          "perPage": 10,
        },
      },
    },
  ],
  "current": Object {
    "tableProps": Object {
      "cells": Array [
        Object {
          "props": Object {
            "width": 35,
          },
          "sortByProp": "title",
          "title": "Task",
          "transforms": Array [
            [Function],
          ],
        },
        Object {
          "props": Object {
            "width": 20,
          },
          "renderExport": [Function],
          "sortByProp": "system_count",
          "title": "Systems",
          "transforms": Array [
            [Function],
          ],
        },
        Object {
          "props": Object {
            "width": 20,
          },
          "sortByProp": "run_date_time",
          "title": "Run date/time",
          "transforms": Array [
            [Function],
          ],
        },
      ],
      "onSort": [Function],
      "rows": Array [
        Array [
          Object {
            "title": "taskA",
          },
          Object {
            "title": 10,
          },
          Object {
            "title": "Running",
          },
        ],
        Array [
          Object {
            "title": "taskB",
          },
          Object {
            "title": 5,
          },
          Object {
            "title": "20 Apr 2022, 11:10 UTC",
          },
        ],
        Array [
          Object {
            "title": "taskB",
          },
          Object {
            "title": 3,
          },
          Object {
            "title": "25 Apr 2022, 11:10 UTC",
          },
        ],
      ],
      "sortBy": Object {
        "direction": "asc",
        "index": 0,
      },
    },
    "toolbarProps": Object {
      "pagination": Object {
        "itemCount": 3,
        "onPerPageSelect": [Function],
        "onSetPage": [Function],
        "page": 1,
        "perPage": 10,
      },
    },
  },
  "error": undefined,
}
`;
